neps/nep-0605-sharded-contracts.md

```diff
@@ -0,0 +1,182 @@
+Type: Protocol
+Version: 0.0.0
+Created: 2025-04-07
+LastUpdated: 2025-04-07
+---
+
+## Summary
+
+Today, a single contract is limited to the transactions per second throughput (TPS) of a single a shard, ergo a contract already at the TPS limit of a single shard cannot benefit from increase in TPS made possible by increasing the number of shards.  This NEP builds on top of the global contracts NEPs to enable sharded contracts.  Sharded contracts will be able to seamlessly scale to use the entire TPS capacity of the network.
+
+## Motivation
+
+As a single contract is deployed on a single shard, the maximum TPS that it can have is the maximum TPS of the single shard.  Horizontally scaling (i.e. increasing the number of shards) a blockchain is easier than vertically scaling (i.e. increasing the TPS of a single shard).  Once, all the software bottlenecks are addressed, the only way to vertically scale a shard is by requiring the validators to use faster machines.  Faster machines are more expensive and thereby hurts decentralisation.
+
+Without any additional primitives at the protocol level, a single contract will therefore remain bound by the TPS throughput of a single shard regardless of how many shards are added to the network.
+
+This NEP proposes solving this problem by introducing some new protocol level primitives which allow a single contract to scale to use the throughput capacity of not just one but all shards of the network.
+
+## Specification
+
+### Background
+
+We will use the fungible tokens (FT) contract as an example contract to explain the specification.  This section will briefly explain how this contract works on the network today.
+
+The contract consists of state where all the user' account balances are stored in a single HashMap data structure.  When a user wishes to transfer some FT to another user, the following steps take place:
+
+- The sender sends a transaction with a function call action to the contract to transfer the tokens from one account to another.
+- This transaction is routed to the shard where the user's account lives where it is converted to a receipt.
+- Then the receipt is routed to the shard where the FT contract lives.
+- Once it arrives on the FT contract's shard, the receipt is executed, the function call is performed, and the transfer is performed.
+
+In the lucky case, the sender's account resides on the same shard as the FT contract.  In this case, depending on the load on the network, the transaction may be converted into a receipt and the receipt is executed in the same block.  Assuming a uniform distribution of user accounts across the network, such cases will only happen for every `N`th transaction where `N` is the number of shards on the network.  In other cases, the minimum latency is 2 blocks.  In the first block, the transaction is converted into a receipt and transferred to the receiver's shard.  And in the second block, the receipt is processed.
```

Comment:
> In the lucky case, the sender's account resides on the same shard as the FT contract. In this case, depending on the load on the network, the transaction may be converted into a receipt and the receipt is executed in the same block.

Not true. The user account is different from the ft contract account, so even if they are on the same shard, the receipt to call ft contract will *not* be executed in the same block

Reply:
Yes, reading https://github.com/near/nearcore/blob/ed1a56d4d33436c59ad1415b1722518ce34bfee2/runtime/runtime/src/lib.rs#L1529 more carefully, I see you are right.  I will fix this comment.